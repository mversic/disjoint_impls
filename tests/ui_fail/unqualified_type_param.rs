use disjoint_impls::disjoint_impls;

pub trait Dispatch {
    type Group;
}

pub enum GroupA {}
pub enum GroupB {}

impl Dispatch for String {
    type Group = i32;
}
impl<T> Dispatch for Vec<T> {
    type Group = u32;
}
impl Dispatch for i32 {
    type Group = GroupA;
}
impl Dispatch for u32 {
    type Group = GroupB;
}

disjoint_impls! {
    pub trait Kita {
        const NAME: &'static str;
    }

    impl<T: Dispatch> Kita for T
    where
        T::Group: Dispatch<Group = GroupA>,
    {
        const NAME: &'static str = "Blanket A";
    }

    impl<T: Dispatch> Kita for T
    where
        T::Group: Dispatch<Group = GroupB>,
    {
        const NAME: &'static str = "Blanket B";
    }
}

fn main() {}
