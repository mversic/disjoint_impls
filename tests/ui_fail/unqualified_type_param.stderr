error: Ambiguous associated type. Qualify with a trait to disambiguate (e.g. < T as Trait > :: Group)
  --> tests/ui_fail/unqualified_type_param.rs:30:9
   |
30 |         T::Group: Dispatch<Group = GroupA>,
   |         ^^^^^^^^
